addon_service/addon_operation_invocation/models.py

```diff
@@ -40,6 +41,24 @@ class Meta:
     class JSONAPIMeta:
         resource_name = "addon-operation-invocations"
 
+    @property
+    def account(self):
+        if not self.thru_account:
+            return None
+        try:
+            return self.thru_account.authorizedstorageaccount
+        except AuthorizedStorageAccount.DoesNotExist:
+            return self.thru_account.authorizedcitationaccount
+
+    @property
+    def addon(self):
+        if not self.thru_addon:
+            return None
+        try:
+            return self.thru_addon.configuredstorageaddon
+        except ConfiguredStorageAddon.DoesNotExist:
+            return self.thru_addon.configuredcitationaddon
```

Comment:
If we need to access shared attributes, we can access them from `thru_addon` or `thru_account` properties

Reply:
yeah, exactly -- i'd suggest removing these `addon` and `account` properties, since they may mislead coders to incur extra cost by accident, using them to access shared attributes instead of `thru_addon` and `thru_account` ([for example](https://github.com/CenterForOpenScience/gravyvalet/pull/104/files/3af702fd3246db402b11c938ecabe8e39168d175#diff-4e286204f143a44396301220ac4e27d9e1b6a1872709b0a103fd5b792056e84eR88) `self.account.imp_cls`)

Reply:
but this way accessing non shared properties (and there are much because we sacrificed DRYness for readablility) would be a nightmare, and would require coder to know whether he operates with `configuredstorageaddon` or `configuredcitationaddon`, which may not be available in the scope

Reply:
if it's a non-shared property, the coder must already know whether `storage` or `citation` or another addon type, because the property is defined on the type-specific subclass -- one of the benefits of this multi-table inheritance approach (as i thought you pointed out) was to DRY these models, undo that sacrifice, so anything that makes sense across addon types can be moved to the base `ConfiguredAddon`/`AuthorizedAccount`/`ExternalService` models

Reply:
I agree, but this would applies only to perfect world (or very complex implementation), because in our cases most of properties are not shared because of `related_name` field, which must be explicitly passed to Fields in our models, and which must be different for all subclasses

Reply:
i'm not sure i understand -- if a relation is defined on a non-abstract base model (using multi-table inheritance), `related_name` can reference the name of that base class and be the same for all subclasses

```py
# if non-abstract base `ConfiguredAddon` has
base_account = models.ForeignKey('AuthorizedAccount', related_name='configured_addons', ...)

# then that relation is available on `AuthorizedAccount` subclasses as well
_storage_account: AuthorizedStorageAccount
for _addon in _storage_account.configured_addons.all():
    _addon.shared_attr  # queryset returns `ConfiguredAddon` instances

# in storage-specific code, where you know you'll need to narrow the type of every instance,
# can use `select_related` to avoid extra queries
for _addon in _storage_account.configured_addons.select_related('configuredstorageaddon'):
    _addon.configuredstorageaddon.storage_specific_attr
```
my sense is that consolidating the many duplicated relations onto base classes would significantly simplify the current implementation

(or, when using abstract base models, [django offers](https://docs.djangoproject.com/en/4.2/topics/db/models/#abstract-related-name) `%(app_label)s` and `%(class)s` related_name placeholders -- if multi-table inheritance is truly untenable, i'd recommend at least replacing the many duplicated relation fields (`oauth2_token_metadata`, `_credentials`, `account_owner`, etc) with relations on the abstract base with placeholder-ed `related_name`)